### ISSUE TITLE
feature request: stub and mock support

### ISSUE THREAD

**Issue by magwas** [xspec/xspec#761]
As a developer I would like to have the ability to provide mocks and stubs for testing.

There is an xslt which uses functions and modes defined in other stylesheets. To able to test it, I need some of the following features:

1. simple import
  One or more xslt is imported in the xspec file. These xslts contain the stub implementation of the functions/modes the sheet under test depends on

2. mocking support
  A simple framework which describes the output to be generated for each test input which a function/mode should generate. The description could be put to the xspec file directly, or imported to it.

Are there already existing features which implement or could support this kind of functionality?
If I want to implement it, where should I start?

**Comment by AirQuick** [comment]
> 2. mocking support
>    A simple framework which describes the output to be generated for each test input which a function/mode should generate. The description could be put to the xspec file directly, or imported to it.

This repo has [`find-examples.*`](https://github.com/xspec/xspec/tree/master/misc/find-examples). Does it resemble your goal to some extent?

**Comment by AirQuick** [comment]
> 1. simple import
>    One or more xslt is imported in the xspec file. These xslts contain the stub implementation of the functions/modes the sheet under test depends on

I often create a set of files like this:

**test/module1.xspec**
```xspec
<x:description stylesheet="module1.xsl">
   <!-- Test scenarios for the functions and templates in ../src/module1.xsl -->
</x:description>
```

**test/module1.xsl**
```xslt
<xsl:stylesheet>
    <!-- external dependencies -->
    <xsl:include href="../lib/functx.xsl" />

    <!-- internal dependencies -->
    <xsl:include href="../src/module2.xsl" />

    <!-- test target -->
    <xsl:include href="../src/module1.xsl" />

    <!-- helper functions used only for running tests -->
    <xsl:function>...</xsl:function>
    ...
</xsl:stylesheet>
```

**src/module1.xsl**
```xslt
<xsl:stylesheet>
    <!-- XSLT module file. Functions and templates in this module
         depend on ../lib/functx.xsl and module2.xsl -->
</xsl:stylesheet>
```

What you're seeking is more advanced or entirely different than this?

**Comment by magwas** [comment]
> This repo has [`find-examples.*`](https://github.com/xspec/xspec/tree/master/misc/find-examples). Does it resemble your goal to some extent?

I cannot see how find-examples would be applicable for this.

**Comment by magwas** [comment]
> I often create a set of files like this:
> 
> **test/module1.xspec**
> 

Yes, this covers the simple import case as done by hand, and it is a useable workaround for the problem.

I would like to have something along the lines

test/module1.xspec

    <x:description stylesheet="module1.xsl">
       <x:import stubs="dependencies1.stubs">
       <x:import stylesheet="dependencies2.xsl">
       
    </x:description>

dependencies1.stubs

    <x:stubs>
       
       <xsl:function name="my:function">
           <xsl:param name="firstParam"/>
         
      </xsl:function>

      
      <x:stub type="function" name="my:function">
          <x:param name="firstParam"/>
          <x:case when="$firstParam = 'foo'" select="'bar'"/>
          <x:case when="$firstParam = 'bar'>foo</x:case>
      </x:stub>

       
         <xsl:template match="*" mode="dependencyMode">
            
        </xsl:template>

       
      <x:stub type="template" match="*" mode="dependencyMode">
          <x:case when="name() = 'foo'"><foo bar="baz"/></x:case>
          <x:case when="name() = 'bar' select="."/>
      </x:stub>

      </x:stubs>

**Comment by AirQuick** [comment]
> 2. mocking support
>    A simple framework which describes the output to be generated for each test input which a function/mode should generate. The description could be put to the xspec file directly, or imported to it.

So you're looking for a simple scheme to inject/override/wrap some static or dynamic functions/templates when running XSpec tests, in order to mimic/verify data flowing through them?

**Comment by magwas** [comment]
In the meantime I have explored this further. I intend to contribute code. Before that I would like to make sure that we implement something which is useful to others and gets accepted.

I guess [this article](https://www.linkedin.com/pulse/test-driven-development-xslt-magos%C3%A1nyi-%C3%A1rp%C3%A1d/) summarizes our way of thought. I have quoted the most important parts below.

Please advice about the way to implement it and your preferences.

> For stubs and mocks we use plain old xslt templates, utilizing xsl:choose based on the input cases. To be able to include them, we test not the template under test itself, but a wrapper which imports the stubs and the template under test. Our implementation pattern also calls for separating test data from test code. We just forgot about it for now, using the already existing x:context construct.
> 
> Of course that means some boilerplate which should not be written in the first place, but this is just meant as an exploration on what kind of capabilities we should contribute to xspec.
> 
> For the stubs, a simple xml structure describing what output should be emitted for what input will do. We just have to add the construct referencing it to the xspec descriptor. As xspec already works by compiling the xspec descriptor to a template file, it is relatively straightforward to do.
> 
> While we are at it, we also intend to implement some ways to refer to test data in external file, making it reusable across tests.
>

**Comment by AirQuick** [comment]
If what you're aiming for has a clear advantage than a [generic method](https://github.com/xspec/xspec/issues/761#issuecomment-577112839), you may want to start from a *tiny* proof of concept and let us know.
See [*Under the hood*](https://github.com/xspec/xspec/tree/master/tutorial/under-the-hood) for how XSpec test suites are compiled.

Please note that even if the implementation looks fine, before accepting it we will generally need these:
* [Schema](https://github.com/xspec/xspec/blob/master/src/schemas/xspec.rnc)
* [Test](https://github.com/xspec/xspec/wiki/How-to-Run-the-Test-Suite-Locally)
* XQuery support (not necessarily on a parity with XSLT support, but at least clear distinction between what is possible and what is impossible)
  (I don't think XQuery is particularly important. But once we leave XQuery behind XSLT, recovering it later without the original developer is very difficult. So, I'm just making sure not to forget XQuery completely.)
